### ISSUE TITLE
Quarkus Rest empty @QueryParam is handled differently

### ISSUE THREAD

**Issue by ia3andy** [quarkusio/quarkus#44885]
### Describe the bug

I initially discovered this in code.quarkus unit tests. For example when using `?a=` with a a List, the list size is not 0 while it was 1 before (with `""`) as content. 

Same with String, before it was an empty string, not it's not defined.

### Expected behavior

not sure but we need to clarify if it's an expected change

### Actual behavior

it assumes `?a=` is the same as `?`

### How to Reproduce?

[reproducer-validation.zip](https://github.com/user-attachments/files/17989638/reproducer-empty-queryparam.zip)

This reproducer works on 3.17

### Quarkus version or git rev

main (3.18)

**Comment by ia3andy** [comment]
It seems the change is not in validation but in how the empty query param is treated
Before (in 3.17-), `?a=` meant `a=""`, in main it's the same as not having a
This makes the validation constraint fail since there is a default value.

I have another failing test on a similar case where `?e=` and e is a list before it was a `List.of("")` and now it's a `List.of()`.

**Comment by ia3andy** [comment]
THis started to fail on Nov 18 and was working on Nov 15

**Comment by ia3andy** [comment]
here is a new reproducer clearly showing that empty query params are now handled differently:

 [reproducer-empty-queryparam.zip](https://github.com/user-attachments/files/17989638/reproducer-empty-queryparam.zip)

**Comment by FroMage** [comment]
This change was intentional, and explained in https://github.com/quarkusio/quarkus/pull/42468#issuecomment-2302312141

If you have use-cases that require the old behaviour, let us know, because we could not think of any.

**Comment by geoand** [comment]
Thanks for spotting that @FroMage!

I had completely forgot about that one...

**Comment by ia3andy** [comment]
@FroMage let's close this when this is in the migraition guide

**Comment by ia3andy** [comment]
Thanks @FroMage @geoand @cescoffier

**Comment by Saphri** [comment]
I maintain a huge old legacy system... we actually use the empty string to signal something different than a null.

Think of this as a form of mutation that is submittet... null means no processing required here... empty string means delete it.. thats the logic that the system has implemented 😄 

How this is totally broken with 3.18.x

**Comment by gsmet** [comment]
I had a look at what initially triggered the issue. From what I could see, it wasn't related to strings and then the fix affected strings to improve consistency.

I agree it's not exactly easy to find a consistent way of doing things but I must admit I'm a bit worried about this change for strings.
Do we know how other REST layers behave? RESTEasy Classic and Spring?

**Comment by Saphri** [comment]
Take these two examples:

path?myQueryParam=&anotherParam=foo
and
path?anotherParam=foo

For me there is a big difference in those two urls, with the first having the intention of submitting a value for that param. So those should not be null both (and IMO any @Default should only kick in on the second one)

**Comment by FroMage** [comment]
The original behaviour was mostly unintentional, non-specified, and not consistent.

You can still obtain the information that the query included one key with no value from the Vert.x request.

If there are multiple users who want to obtain this information, we could add a new API as a new feature to get that info, but so far I think this is the first time someone asks for that.

**Comment by Saphri** [comment]
Using that Vert.x request as a workaround might be easy enough. But finding all places, to use this workaround, that now breaks in 100's of endpoints will be a major pain here

**Comment by FroMage** [comment]
Sorry :(

You could also turn these `String` parameters into a new type `OptionalString` via a `ParamConverter`, and make that type have `set/empty/not-set` states. But that would still mean rewriting the endpoints, so not much more helpful.

**Comment by FroMage** [comment]
Or… perhaps you can write a request filter that turns `foo=` into `foo=<marker>` and then you can refactor your calls from `StringUtil.isEmpty(param)` into one that treats `<marker>` specially?

**Comment by Saphri** [comment]
Would it be possible with a config to enable legacy behavior and some kind of log message that we can look for and fix? I much rather fix the code correct than doing lots of workarounds.. and just make 3.19.x the cutoff?

**Comment by Saphri** [comment]
A flag to enable legacy behaviour, in itself, would not help us find all the places where this impacts us.

Any time a param would be null and not empty string, it should emit a log message. "<param> will be null with new behaviour on endpoint <x>". We would still need to analyse the exact impact on that endpoint, but it will likely reduce the number of endpoints greatly that we have to go over

**Comment by FroMage** [comment]
So let me rephrase this: are you looking for a build-time warning whenever we see a `String` parameter, or a runtime warning when we see an empty string parameter value?

**Comment by Saphri** [comment]
When we encounter query parameters like this:
path?myQueryParam=&anotherParam=foo

It could issue a message that myQueryParam is no longer an empty string but null.. at runtime yes
This will help alot with finding all the places where this breaking change with query parameters have an effect at runtime. Because one of the big issues with this breaking change, is that the compiler cannot catch this change... so you get the suprise at runtime.. likely as a production issue for future me 😄 

Now, I think, there should be a flag to enable this, so its not spamming these messages at default

**Comment by michieldondorp** [comment]
We hit the same issue, I dislike the option was removed and I have no (in my opinion) clean way to replace the old way of working.

We used to filter by group, in which we had 3 options (3&4 are basically the same):
* no filter -> omitting of queryparam
* empty list, exclude all that are grouped -> queryParam=
* filter single -> queryParam=x
* filter multiple -> queryParam=x&queryParam=y

Im no longer able to see the difference between omitting the queryParameter and adding an empty queryParameter

Basically I have 3 options:
- Add another parameter for exclude all grouped
- Check for the null in queryParameters
- Use a comma-separated-string and regain the option for null

I would love to have the an implementation for @DefaultValue(null), allowing to regain control over this (undocumented) sideeffect.

**Comment by olivierbeltrandocintoo** [comment]
Hello, hello,

> If you have use-cases that require the old behaviour, let us know, because we could not think of any.

Here is a proposed usecase :)
A queryParam that is a list of included types for an API. ex: fileTypes.

here would be an API consumer use case.
```kotlin
val myTypes = computeRequiredTypes()

getFiles(myTypes) // API call to /api/files?fileTypes=
	.processFiles()  // process those files
```

Now imagine that the `computeRequiredTypes()` would return an empty list.

On the old behaviour `getFiles()` would return an empty list, which would play nicely in a logic chain of functionnal programming. Of course the getFiles(myTypes) would return an empty list, and a network call could have been avoided but still logic is not broken.

On the new behaviour `getFiles(fileTypes = null)` would return all files, and processFiles() would have to treat them.

From this perspective, to me, it looks like an inferior default with a riskier chance of traps.

About http, the QueryParams are strings, it's the only semantic type there is really.

We use Quarkus to serve our APIs to our customers, and as API provider we try to provide the most stability, but this litterally shifts the ground under our feets.

Not trying to freeze everything in the rock here, but I kinda hope you will reconsider.

Thanks,
Cheers Olivier

**Comment by FroMage** [comment]
> When we encounter query parameters like this: path?myQueryParam=&anotherParam=foo
> 
> It could issue a message that myQueryParam is no longer an empty string but null.. at runtime yes This will help alot with finding all the places where this breaking change with query parameters have an effect at runtime. Because one of the big issues with this breaking change, is that the compiler cannot catch this change... so you get the suprise at runtime.. likely as a production issue for future me 😄
> 
> Now, I think, there should be a flag to enable this, so its not spamming these messages at default

OK, thanks for the clarification.

**Comment by ia3andy** [comment]
The more I look at it the more I think it was a  mistake to remove this behaviour. We should at least provide a way to mark if which behaviour we want to use for a given param.

**Comment by FroMage** [comment]
> Hello, hello [@FroMage](https://github.com/FroMage),
> 
> > If you have use-cases that require the old behaviour, let us know, because we could not think of any.
> 
> Here is a proposed usecase :) A queryParam that is a list of included types for an API. ex: fileTypes.
> 
> here would be an API consumer use case.
> 
> val myTypes = computeRequiredTypes()
> 
> getFiles(myTypes) // API call to /api/files?fileTypes=
> 	.processFiles()  // process those files
> Now imagine that the `computeRequiredTypes()` would return an empty list.

I'm sorry I'm not following, are we talking about how the client translates an empty list of parameters, or the server?

I think the client would translate `getFiles(myTypes)` to `/api/files` for an empty list. It would have to be the very special case of a 1-element list containing an empty string to translate it to `/api/files?fileTypes=` which you'll have to agree is an extremely special-casey value to denote "set this to something empty on the server, but not `null`", no?

**Comment by FroMage** [comment]
People, we did not expect so many users to be affected by this, but you are, and that's not good. We want to fix this.

However, I'm still not seeing how the previous behaviour was logical in any way, _especially_ in the case of `foo=` generating a 1-element list containing an empty string. This is a very weird way to specify that the value is specified but empty.

Just to summarise things as they stand:

| Quarkus Version | Parameter Type | HTTP value | Deserialised value |
| --- | --- | --- | ---
| 3.18 | `String` | <nothing> | `null`|
|  |  | foo= | `null`|
|  |  | foo=a | `"a"`|
|  |  | foo=a&foo=b | `"a"`|
|  | `List<String>` | <nothing> | `[]` (empty list) |
|  |  | foo= | `[]` (empty list) |
|  |  | foo=&foo= | `[]` (empty list) |
|  |  | foo=a | `["a"]` (1 element list) |
|  |  | foo=a&foo=b | `["a", "b"]` (2 element list) |
| 3.17 | `String` | <nothing> | `null`|
|  |  | foo= | `""` (empty string) |
|  |  | foo=a | `"a"`|
|  |  | foo=a&foo=b | `"a"`|
|  | `List<String>` | <nothing> | `[]` (empty list) |
|  |  | foo= | `[""]` (1 element list with empty string) |
|  |  | foo=&foo= | `["", ""]` (2 element list with empty strings) |
|  |  | foo=a | `["a"]` (1 element list) |
|  |  | foo=a&foo=b | `["a", "b"]` (2 element list) |

I agree we need a solution for you, though.

We have several options:

- Revert the previous behaviour (even though it is inconsistent and special-casey) and justify it by saying that empty strings really have different semantics to `null`, even though this is a really hard thing to justify
- Revert the previous behaviour under a configuration option that defaults to `false`
- Provide a new API to allow distinguishing three states: absent, present and empty, present and non-empty. This would work with not just `String` but any type, unlike the old behaviour.
- Revert the previous behaviour for single-valued strings, pretending that empty strings are different to `null`, but for multi-valued parameters, change `<nothing>` into `null` (as opposed to an empty collection), so that `foo=` can be `[] (empty collection)`, allowing you to get rid of the horrible `[""] (list of empty string)`, even though that'd still be irregular.

The new API could be something like:

```java
public record Parameter<T>(T value, boolean specified){}
```

This would allow you you use `@RestQuery Parameter<String> foo` and differenciate wether they're absent (unspecified) or present (specified) and this even for non-`String` types, which have no empty value. For those, the empty specified value would be `null`, just as the unspecified value, but you can use the `boolean` to differentiate.

Similarly this would work for `@RestQuery Parameter<List<String>> foo` and allow you to differentiate the `foo=` case without resorting to a 1-element list containing an empty value.

Think about these options, and let us know which ones you prefer, which are acceptable, or wether you have any better idea?

We're going to find a solution, bear with us.

**Comment by FroMage** [comment]
I'm reopening this issue.

**Comment by michieldondorp** [comment]
> 
> > * Use a comma-separated-string and regain the option for null
> 
> Can you explain how this would be different? I'm not following.

My assumption here was that using a string parameter allowed the empty case, the supplied table invalidates this case (my assumption was wrong). 

> > I would love to have the an implementation for [@DefaultValue](https://github.com/DefaultValue)(null), allowing to regain control over this (undocumented) sideeffect.
> 
> Excuse me, I'm not following this either: we can't specify `@DefaultValue(null)` because `null` is not valid in an annotation. Can you rephrase this?
> 
Using `@DefaultValue` with nullvalue is not possible for you cant use null for you need to supply a constant (https://docs.oracle.com/javase/specs/jls/se8/html/jls-9.html#jls-9.7.1) and null is not a valid constant: (https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.28)


> > The documentation of the spec is nonconclusive, the spec for [@DefaultValue](https://github.com/DefaultValue) is conclusive though: https://javaee.github.io/javaee-spec/javadocs/javax/ws/rs/DefaultValue.html
> > ```
> > If this annotation is not used and the corresponding meta-data is not present in the request, the value will be an empty collection for List, Set or SortedSet, null for other object types, and the Java-defined default for primitive types. 
> > ```
> 
> Here as well, I'm not following what you mean. This is coherent with our current behaviour, as far as I can see.

Yes, this comment is to agree with the choice to be consistent with the current behavior. From my perspective these use cases are nice to haves, for we have alternatives and the specification is somewhat clear (definitely in the documentation of DefaultValue, less so in the javaee spec). Perhaps an optional parameter for `@DefaultValue` to differentiate between the 'omitted' and 'empty' parameter states.

**Comment by FroMage** [comment]
This seems logical… until you consider `foo=&foo=` in the collection case: should it still be `[]` (empty)? I also have to raise that it's entirely arbitrary that `foo=` means "empty collection" more than "null value" or even the current situation of "collection of single empty string".

Also, how does this transpose to primitives or even other types that have no empty value, such as `Date` or `Integer`?

In the case of `Date` (single), both `` (absent) and `foo=` (present, empty) would yield `null` and we still can't differentiate between absent or empty.

Last, not that while I agree it makes sense to have `null` for `List<*>` when there are no parameters, I'm 100% sure that people would complain if we made the move now.

**Comment by geoand** [comment]
@FroMage thanks so much for being thorough on this.

Given how weird the whole mapping is, my proposal is that we stay with what we have now (as it's at least consistent), but that we also provide the `Parameter` API you mentioned.

**Comment by FroMage** [comment]
That would be my suggestion as well, but I haven't heard from people here that this would work for them?

**Comment by olivierbeltrandocintoo** [comment]
Hello, hello

> That would be my suggestion as well, but I haven't heard from people here that this would work for them?

Sorry for that, and also sorry for the time to answer. I really appreciate your involvement in this.

> * Provide a new API to allow distinguishing three states: absent, present and empty, present and non-empty. This would work with not just String but any type, unlike the old behaviour.
> * Revert the previous behaviour for single-valued strings, pretending that empty strings are different to null, but for multi-valued parameters, change <nothing> into null (as opposed to an empty collection), so that foo= can be [] (empty collection), allowing you to get rid of the horrible [""] (list of empty string), even though that'd still be irregular.

Both those options look good to me, especially the one that get rid of the horrible [""].

I am going to repeat over what we all said above "" and null are different and are distinct semantics as much as `foo=` is different than `nothing`.
Note for instance in Javascript there is `undefined` and `null` that matches that semantic well.
In Typed languages (like Java,Kotlin,...) it's a whole different story as an object has a predetermined set of attributes which eliminates the need for this distinction.
But **for the web**, that semantic exist and should be preserved.

The more I think of it, and the more I think because of the format of query parameters (text based key value) that it actually is a ***Map<String, String>*** (in Java/Kotlin) / or a json with key strings (obviously! same limitation of string) and values only of type string.

So, I am thinking that the table should in reality look like this (allow me to add not null types)

|Parameter type| HTTP value | Deserialized value|
| ------------- | ------------- | ------------- |
| `String`        |               | `null`  |
|                 | foo=          | `""` // empty semantic |
|                 | foo=a         | `"a"`  |
|                 | foo=a&foo=b   | `"b"`  // last put in the map overwrites|
| `String!`       |               | throw  |
| `List<String>`  |               | `null`  |
|                 | foo=          | `[]` // empty semantic |
|                 | foo=a         | `["a"]`  |
|                 | foo=a&foo=b   | `["b"]` // last put in the map overwrites |
|                 | foo=a,b   | `["a","b"]` // here allow customtization for separator  |
|                 | foo=,   | `["",""]` // "universal" result of doing a split by comma on `,`   |
| `List<String>!` |               | throw  |

The `,` Separator is across languages **the** separator.
-> java, python, kotlin, rust, haskell, you name it, the comma separator separates arguments, values.
So it could be a good idea to use it as a default to parse a text that is supposed to have a list semantic.

Note1 about ',': that it can already be customized with the `@Separator(",")` annotation.
Note2 about ',': on python, nodeJs, kotlin applying a `split` on `,` using the comma separator all return `["",""]`. This is pretty well accepted :) 

![Image](https://github.com/user-attachments/assets/53325701-5dac-40fc-8292-85e3e43a03ba)
![Image](https://github.com/user-attachments/assets/c815c6b3-6526-4865-937e-4bffa47b6009)
![Image](https://github.com/user-attachments/assets/c69ee522-a849-48e4-8452-026744b49dde)

About "primitives" or actually any other kind of data than string, it involves parsing with the rules you get to define by the framework, and allow customization or not. You are fully right this is arbitraty. 
This is how I would see it. If I were a JsonSchema validator and received the following.

|Parameter type| HTTP value | Deserialized value|
| ------------- | ------------- | ------------- |
| `Int`        |               | `null`  |
|                 | foo=          | `throw 400 response`|
|                 | foo=1         | `1`  |
|                 | foo=a         | `throw 400 response`  |
|                 | foo=1&foo=2   | `2`  |
| `Int!`        |               | throw |
| `List<Int>`  |               | `null`  |
|                 | foo=          | `[]` // empty semantic |
|                 | foo=1         | `[1]`  |
|                 | foo=a         | `throw 400 response`  |
|                 | foo=1&foo=2   | `[2]`  |
|                 | foo=1,2   | `[1,2]`  |
| `List<Int>!`    |               | throw |


So even if `0` could be seen as a nice default, then we could ask ourselves why not `1` ? -> then no just throw a 400. it's not a number.
Same for the date, same for everything actually. 
The API states that it accepts a `double` -> if it's not a double, then let's throw, unless the coder specified that it wants a specific default.

Again, **both of the options proposed work** :) Flexibility is always welcome :)
Consider the rest of the message as matter of though for making *good* default, although as you said everything is arbitraty. 

Sorry for the long message 

**And again thanks a lot** !! 
Cheers,
Olivier

**Comment by olivierbeltrandocintoo** [comment]
Hello hello, do you have any news ? 
We are moving forward with a custom parsing of the UriInfo for now to capture the initial input while upgrading to Quarkus 3.19.3
Thanks

**Comment by geoand** [comment]
@FroMage is this one on your radar? Ideally we'll need a solution we can backport to 3.20 because of the breakage between LTS versions

**Comment by FroMage** [comment]
> Hello hello, do you have any news ? We are moving forward with a custom parsing of the UriInfo for now to capture the initial input while upgrading to Quarkus 3.19.3 Thanks

Naturally, you can already use this:

```java
@GET
public void foo(@RestQuery String single, @RestQuery List<String> multiple, UriInfo uriInfo){
 if(single != null) {
  // it must be set to a value
 } else if(uriInfo.getQueryParameters().containsKey("single")) {
  // cleared (passed as `single=`)
 }
 if(!multiple.isEmpty()) {
  // it must be set to at least one value
 } else if(uriInfo.getQueryParameters().containsKey("multiple")) {
  // cleared (passed as `multiple=`)
 }
}
```

We could have our own subtype of `UriInfo` to make this easier:

```java
@GET
public void foo(@RestQuery String single, @RestQuery List<String> multiple, RestUriInfo uriInfo){
 if(single != null) {
  // it must be set to a value
 } else if(uriInfo.isQueryCleared("single")) {
  // cleared (passed as `single=`)
 }
 if(!multiple.isEmpty()) {
  // it must be set to at least one value
 } else if(uriInfo.isQueryCleared("multiple")) {
  // cleared (passed as `multiple=`)
 }
}
```

But that's only marginally better. I think I prefer the `Parameter<T>` API.

**Comment by FroMage** [comment]
Yes. When is that again?

**Comment by geoand** [comment]
https://github.com/quarkusio/quarkus/wiki/Release-Planning#320-lts---branched-from-319-proposal 

The window for the first release has already passed, but we really need something we can backport

**Comment by ia3andy** [comment]
I am leaning toward:
"Revert the previous behaviour for single-valued strings, pretending that empty strings are different to null, but for multi-valued parameters, change <nothing> into null (as opposed to an empty collection), so that foo= can be [] (empty collection), allowing you to get rid of the horrible [""] (list of empty string), even though that'd still be irregular."

**Comment by ia3andy** [comment]
Only issue is that with this change, we would have 3 behaviours 🤯

**Comment by geoand** [comment]
I am -1 on reverting as that will just increase the confusion.

**Comment by ia3andy** [comment]
> I am -1 on reverting as that will just increase the confusion. I like the `Parameter` idea a lot we just need to have it soon :)

It's not 100% reverting it's reverting to something that solves both issues (single valued can detect empty, multi value too). @FroMage reverting was probably not the right way to mention this solution.

The `Parameter` solution is fine but heavier.

**Comment by FroMage** [comment]
> The window for the first release has already passed, but we really need something we can backport

There's no way I can do `Parameter` properly before tomorrow's `Final`.

At this point, if we want to backport something before tomorrow, the only option seems to be a config option to put the previous behaviour back. And I'm not entirely sure that's really required, given that with `UriInfo` there's an alternative.

The alternative requires updating your code, but so does using the new `Parameter` API.

I'm not sure what to do here.

**Comment by geoand** [comment]
I don't think anyone will object to adding Parameter for .1 as it won't be breaking and it gives a way forward for users moving from one LTS to another

**Comment by geoand** [comment]
Correct

**Comment by FroMage** [comment]
@olivierbeltrandocintoo: find comments inline in bold:

|Parameter type| HTTP value | Deserialized value|
| ------------- | ------------- | ------------- |
| `String`        |               | `null`  |
|                 | foo=          | `""` // empty semantic **this is the old behaviour** |
|                 | foo=a         | `"a"`  |
|                 | foo=a&foo=b   | `"b"`  // last put in the map overwrites|
| `String!`       |               | throw  |
| `List<String>`  |               | `null`  **this will break a lot of existing code that rely on collections never being `null`** |
|                 | foo=          | `[]` // empty semantic |
|                 | foo=a         | `["a"]`  |
|                 | foo=a&foo=b   | `["b"]` // last put in the map overwrites **This is really not how HTTP works for parameters**|
|                 | foo=a,b   | `["a","b"]` // here allow customtization for separator  |
|                 | foo=,   | `["",""]` // "universal" result of doing a split by comma on `,`   |
| `List<String>!` |               | throw  |

> The `,` Separator is across languages **the** separator.

Not for HTTP parameters, which REST is really standardised on. We do support it, but in addition to the HTTP semantics, not as a replacement for it.

As for your primitive table, I am not sure if you meant `int` (the primitive) or `Integer` (the boxed type).

In any case, deserialising `foo=` to something else than 0 (for primitives) or `null` (for non-`String` types) [which is what's happening ATM] would also break a lot of existing code, and we have to remember we're all here because I broke existing code and we should not fix this by breaking other existing code, especially when the reason is not very convincing. 

Throwing a 400 when the type is incompatible (say `a` for an `int` or `23` for a `Date`) is fine. But throwing a 400 for an empty value (`foo=`) for either a primitive or a non-`String` type would be hard to justify.

**Comment by FroMage** [comment]
Closed by accident.

**Comment by FroMage** [comment]
I'm making progress on this. I have the non-parameter-container version ready already. Hopefully by the end of the week I can get parameter containers to also work. Perhaps.

**Comment by FroMage** [comment]
This is a bit all over the place. I thought I could afford to copy the code for `Optional` since it's also a wrapper type, but it's special-cased to support types like `set/list` but for example, not arrays. So ATM `Optional<String[]>` does not work. I'm not fixing this, will wait for someone else to raise the issue, but the support for wrapper types is not recursive properly in that all the types that are wrapped by it are special-cased :(

**Comment by FroMage** [comment]
> Optional<String[]> 

Nevermind, thinking about this, this type makes no sense to support 😅

**Comment by Sgitario** [comment]
We're hitting the same issue here, but with the over-complication that we can't change the API because it's auto-generated by the OpenApi extension. Let me explain further this use case:

- We have an API spec that uses a Query param which is an Enum type:

```
  /api/rhsm-subscriptions/v1/capacity/products/{product_id}/{metric_id}:
    description: 'Operations for capacity report for a given account and product'
    parameters:
      - name: product_id
        in: path
        required: true
        schema:
          $ref: "#/components/schemas/ProductId"
        description: "The ID for the product we wish to query"
    get:
      summary: "Fetch a capacity report for an account and product."
      operationId: getCapacityReportByMetricId
      parameters:
        - name: usage
          in: query
          schema:
            $ref: '#/components/schemas/UsageType'
          description: "Include only capacity for the specified usage level."
``` 

Where the UsageType is:

```
    UsageType:
      description: "Describes the usage that the report was made against. Not set if it wasn't
          specified as a query parameter."
      type: string
      enum: [ "", Production, Development/Test, Disaster Recovery, _ANY ]
```

The autogenerated API is:

```java
@RegisterRestClient
@RegisterProvider(ApiExceptionMapper.class)
@Path("/api/rhsm-subscriptions/v1/capacity/products/{product_id}/{metric_id}")
public interface CapacityApi  {

    /**
     * Fetch a capacity report for an account and product.
     *
     */
    @GET
    
    @Produces({ "application/vnd.api+json" })
    CapacityReportByMetricId getCapacityReportByMetricId(@PathParam("product_id") ProductId productId, @QueryParam("usage") UsageType usage) throws ApiException, ProcessingException;
}
```

- We are migrating some services that use this API spec from Spring Boot to Quarkus. In Spring Boot, this worked fine perfectly because when the query "usage" is unset, it returns null, and when the query "usage" is set with empty value, it returns the UsageType.EMPTY.
- In Quarkus, it always return null. 

Note that I don't follow how a query param that always returns null regardless if it was unset or empty goes in the direction of being more consistent... (I didn't follow all the conversation though), but unless there is a `workaround` that we could apply for all query params, this is blocking us to migrate our services to Quarkus :(

**Comment by FroMage** [comment]
Thanks @Sgitario, I had not thought about openapi generation and custom parameter converters that actually do support empty values.

I was working hard to complete support for the `Parameter<T>` type which allows differentiating between absent/empty/set values, and I have it working for query parameters. I was going to add support for form parameters too, because that seems equally logical to allow it there (let's deal with cookies, matrix, path and headers in the future), but ran into issues due to multipart having specific code paths (as always). So I would not have enough time to fix this before I go on PTO.

Now, given what you said, I'm inclined to propose a different strategy:

- We revert the changes that led to removing empty values. So we'd be back to `null | "" | "..."` for strings, `[], [""], ["..."]` for collections of strings, and `0 (absent value) | 0 (empty value) | 0 (set value)` for primitives (:puke:) and `null (absent value) | <exception> (empty value) | 2025 (set value)` for dates and other types whose converter does not allow empty values.
- We still introduce `Parameter<T>` for people who want to deal with empty values for any kind of type: string, primitives, dates in a sane and consistent manner
- We document the usage of `Parameter` if anybody wants to deal with empty values, making sure that regular users don't fall into the madness that is the JAX-RS default in those cases

The problem is that this will likely have to wait for my return in three weeks :-/

WDYT @geoand ?

**Comment by geoand** [comment]
> We revert the changes that led to removing empty values

I am not sure this is a good thing to do TBH as I believe it will lead to even more confusion

**Comment by geoand** [comment]
The way I see it, we should just include the new API when we have it (the sooner the better) in a next release of 3.20 LTS and tell people that really rely on the old (undocumented) behavior to wait

**Comment by FroMage** [comment]
I see your point. That's still a possibility.

However, I don't know how we can deal with openapi generators, it's likely they won't be able to target this new API. Although it depends on whether openapi can distinguish element types and whether they allow for empty values or not.

I used to think only `String` allowed empty values (as opposed to say primitives, or date/time) but the example with a custom parameter converter that does support empty values was not one I had thought about.

With the current behaviour, and even the addition of the `Parameter<T>` API, this sort of custom parameter converter won't be possible, because:

- With the current behaviour, we remove empty values, so the parameter converter isn't called
- With the `Parameter<CustomParam>` API, the framework will detect the empty parameter and make a special `Parameter.empty()` instance which won't call the parameter converter. So the fact that it's empty is actionable by the user. but it's further up the chain. Well, perhaps it doesn't matter too much.

**Comment by geoand** [comment]
I don't see a way to please everyone to be honest and I am -1 for causing even more confusion by flip-flopping

**Comment by geoand** [comment]
> @geoand since the original behaviour changed and there are no alternatives, I see this like a major regression issue. Note that many users might have not noticed this yet, but expect query params to be treated like null and empty values differently.

But there will be soon. 

> Note that many users might have not noticed this yet, but expect query params to be treated like null and empty values differently.

I'm sympathetic to that, but we also had users that expected the exact opposite (which is why we changed it). The fact is that the old behavior was an emergent property of the implementation, not something that was thought out or specified.

**Comment by FroMage** [comment]
> there are no alternatives

There are workarounds, currently. We've documented one here: https://github.com/quarkusio/quarkus/issues/44885#issuecomment-2733701215

> ?name=null might mean “ignore the name filter.” or “use default behavior.”

I assume you mean `…?` (no query param name nor value)

> Changing that, it's about forcing users/devs to start using optionals

I'm not sure what you mean about optionals here, but I don't think it's the `java.lang.Optional` type. I guess you refer to an extra parameter to specify whether you want an empty query parameter as opposed to an absent one.

**Comment by FroMage** [comment]
> I don't see a way to please everyone to be honest and I am -1 for causing even more confusion by flip-flopping

I am sympathetic to that, though I suppose it depends a bit on how this can be adapted for OpenAPI generators, perhaps @phillip-kruger knows if in open-api we can detect that some query parameter types allow for an empty value or not (as opposed to an absent state).

I would rather do as you suggest and move on from the previous behaviour. But I'm genuinely surprised by the number of reactions to this change. I did not expect that. I also did not expect that some parameter converters allowed for empty values.

So, I'm a bit torn between both options. And when torn, I would tend to trust @geoand's opinion. Who knows, perhaps when I come back in three weeks we'll even have a clear path forward :)

**Comment by FroMage** [comment]
> I undestood from the thread, that another solution was to start using java.lang.Optional for the query parameters that needed to differentiate between null and empty.

Ah OK. Well, that's the new `Parameter<T>` API. `Optional` only has two states: empty or set, we're not using `null` values for `Optional` types because that leads to madness. `Parameter<T>` has three states (and is never `null` itself, just like `Optional` values): absent/cleared/set.

You do need a new API, but on the other hand, the previous behaviour:

- did not work for non-`String` types (except the very special case of having a custom parameter converter which did handle empty values) such as primitives or date/time
- led to frankly silly things such as `[""]`: list with one element containing an empty string to indicate "cleared collection"
- was not used by a lot of use-cases: while many people showed up here to complain, far more users did not. The ability to differenciate between a cleared and absent value is not extremely common, hence why nobody even noticed you could not do that for non-`String` types (* minus the exception noted above)

**Comment by dhufnagel** [comment]
> This change was intentional, and explained in [#42468 (comment)](https://github.com/quarkusio/quarkus/pull/42468#issuecomment-2302312141)
> 
> If you have use-cases that require the old behaviour, let us know, because we could not think of any.

I am really surprised and a little upset about intentional breaking changes just because someone "would be surprised if anyone is relying on this behavior". In a project as widely used as quarkus, there will always be someone which relies on non-breaking behaviour.

**Comment by jheyd** [comment]
Just to not be part of the "far more users" that did not complain: This just broke one of our APIs and we now have to check a bunch of others.

I also find it pretty strange that there is now no way to pass empty strings as query parameters when using `@RestQuery`.

A consequence of that is that `@NotNull` for strings now implies `@NotEmpty`/`@NotBlank`, which feels like very surprising behavior to me.

Empty strings and null can have very different semantic meanings and I think the previous behavior was actually less surprising than the current behavior.

**Comment by Sgitario** [comment]
> Empty strings and null can have very different semantic meanings and I think the previous behavior was actually less surprising than the current behavior.

I would like to also note that as it was working before, it's also how works other frameworks like Spring Boot / Micronaut / Resteasy. 

Actually, in my team, to avoid having to modify all our API, the only workaround was to stop using the Quarkus Resteasy Reactive extension, and replace it with Quarkus Resteasy (legacy) extension.

**Comment by zhangjiangqige** [comment]
Our "use case" of the original behavior is that we use custom method scanner and parameter extractor to mimic Spring's behavior of concating values of the same parameter found in both queries and forms, which is because we **have to** maintain Spring compatibility of our codebase and make sure the codebase built with Spring and Quarkus have the same external behavior, possibly for a very long time.

**Comment by geoand** [comment]
@FroMage what happened to the new API you had in the works?

**Comment by FroMage** [comment]
Ooops, I guess I forgot about it when coming back from PTO 😱 I should revive the PR.

**Comment by geoand** [comment]
🙏🏽
